### PR TITLE
msg ROS2 compatibility, microdds_client improvements (timesync, reduced code size, added topics, etc), fastrtps purge

### PR DESCRIPTION
**Pull request by dagar**
- update all msgs to be directly compatible with ROS2
 - microdds_client improvements
   - timesync
   - reduced code size
   - lots of other little changes
   - add to most default builds if we can afford it
   - ~~add new SYS_USB_MODE parameter for optional USB start~~ (I'll revisit this part later)
 - purge fastrtps (I tried to save this multiple times, but kept hitting roadblocks)
 - px4_ros_com https://github.com/PX4/px4_ros_com/pull/157
 - px4_msgs https://github.com/PX4/px4_msgs/pull/15

##### TODO
 - [ ] cleanup and finalize (more generated code reduction, etc
 - [ ] documentation overhaul (because purging fastrtps was unavoidable we need to do a clean sweep with everything working together)

#### Next steps after this PR
 - fully split out PX4 topics from .msg files (https://github.com/PX4/PX4-Autopilot/pull/18911) and build PX4 within colcon
     - at this point we'll no longer need px4_msgs at all, but might want to consider keeping it as an interface package?

### CODE REVIEW

File: src/modules/microdds_client/microdds_client.cpp [hunk 1]
```diff
@@ -442,14 +574,19 @@ MicroddsClient *MicroddsClient::instantiate(int argc, char *argv[])
...
+#endif
 	const char *device = nullptr;
 	const char *ip = "127.0.0.1";
 	int baudrate = 921600;
 	const char *port = "15555";
 	bool localhost_only = false;
+	const char *client_namespace = nullptr;//"px4";
 
-	while ((ch = px4_getopt(argc, argv, "t:d:b:h:p:l", &myoptind, &myoptarg)) != EOF) {
+	while ((ch = px4_getopt(argc, argv, "t:d:b:h:p:l:n", &myoptind, &myoptarg)) != EOF) {
```
Comment (dsix-ls2n) [inline src/modules/microdds_client/microdds_client.cpp:589]: If I am not wrong, the getopt considers options with letters and : are not separator but there to tell if there is an argument with the option.
So it should be` t:d:b:h:p:ln:`

File: src/modules/microdds_client/dds_topics.h.em [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+@[    end for]@
+
+	uint32_t num_payload_sent{};
+
+	void update(uxrSession *session, uxrStreamId create_stream_id, uxrStreamId publish_stream_id, uxrObjectId participant_id, const char *client_namespace);
+};
+
+void SendTopicsSubs::update(uxrSession *session, uxrStreamId create_stream_id, uxrStreamId publish_stream_id, uxrObjectId participant_id, const char *client_namespace)
+{
+	int64_t time_offset_us = session->time_offset / 1000; // ns -> us
```
Comment (dsix-ls2n) [inline src/modules/microdds_client/dds_topics.h.em:44]: There is an issue here with synchronization. The time offset of the session is based on epoch while the timestamps are usually taken from monotonic time. So this shift doesn't work if you want to transform the timestamp between the companion computer and the PX4. A working solution is (maybe not the best):
`
int64_t time_offset_us = uxr_epoch_nanos(session) / 1000 - hrt_absolute_time();
`

File: src/modules/microdds_client/dds_topics.h.em [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+	uint32_t num_payload_received{};
+
+	bool init(uxrSession *session, uxrStreamId stream_id, uxrStreamId input_stream, uxrObjectId participant_id, const char *client_namespace);
+};
+
+static void on_topic_update(uxrSession *session, uxrObjectId object_id, uint16_t request_id, uxrStreamId stream_id,
+		     struct ucdrBuffer *ub, uint16_t length, void *args)
+{
+	RcvTopicsPubs *pubs = (RcvTopicsPubs *)args;
+	const int64_t time_offset_us = session->time_offset / 1000; // ns -> us
```
Comment (dsix-ls2n) [inline src/modules/microdds_client/dds_topics.h.em:84]: Same here
'int64_t time_offset_us = uxr_epoch_nanos(session) / 1000 - hrt_absolute_time();'

File: src/modules/microdds_client/microdds_client.cpp [hunk 4]
```diff
@@ -142,58 +178,73 @@ void MicroddsClient::run()
...
+		// 			      "<dds>"
+		// 			      "<participant>"
+		// 			      "<rtps>"
+		// 			      "<name>default_xrce_participant</name>"
+		// 			      "</rtps>"
+		// 			      "</participant>"
+		// 			      "</dds>" ;
+		// uint16_t participant_req = uxr_buffer_create_participant_xml(&session, reliable_out, participant_id, domain_id, participant_xml, UXR_REPLACE);
+
+		const char *participant_name = "px4_xrce_participant";
```
Comment (dsix-ls2n) [inline src/modules/microdds_client/microdds_client.cpp:245]: I would include the participant namespace here. It is not mandatory for this to work but when you use [micro-ROS-Agent](https://github.com/micro-ROS/micro-ROS-Agent) it creates two nodes with the same name in ROS2 and a lot of tools in ROS2 don't like this.

Somethin uneder the form namespace/px4_xrce_participant.

**Review by dsix-ls2n — COMMENTED** [review]
Hi Daniel, 
Here are some comments from my tests. 
best :)

File: src/modules/microdds_client/dds_topics.h.em (hunk 2 above)
Comment (dagar) [reply src/modules/microdds_client/dds_topics.h.em:44]: Did you see the commit this is carrying within XRCE-DDS to use hrt_absolute_time() for uxr_nanos()? 

https://github.com/PX4/Micro-XRCE-DDS-Client/commit/eff557dddbee5df5afe1da92d121d5f29d0cd006#

Comment (dsix-ls2n) [reply src/modules/microdds_client/dds_topics.h.em:44]: I didn't see that but it is not the version of the submodule in this PR, is it?

**Comment by TSC21** [comment]
@bastianhjaeger @DanMesh FYI

File: src/modules/microdds_client/dds_topics.yaml [hunk 5]
```diff
@@ -0,0 +1,96 @@
+#####
+#
+# This file maps all the topics that are to be used on the micro DDS client.
```
Comment (spiderkeys) [inline src/modules/microdds_client/dds_topics.yaml:3]: Do you have an estimate for how much bandwidth these topics produce in total on your average quadrotor platform?

It looks like just the `sensors_combined`, `vehicle_attitude`, `vehicle_local_position`, and `vehicle_global_position` alone might have the capacity to take up 500K-700Kbits/s, so without rate limiting/filtering, I'm wondering where we stand with this default message set going over a 1MBaud serial interface.

File: src/modules/microdds_client/microdds_client.cpp [hunk 6]
```diff
@@ -46,27 +45,66 @@
...
+				usleep(1'000'000);
+
+			} else {
+				break;
+			}
 		}
 
 		_transport_serial = new uxrSerialTransport();
 
 		if (fd >= 0 && setBaudrate(fd, baudrate) == 0 && _transport_serial) {
```
Comment (spiderkeys) [inline src/modules/microdds_client/microdds_client.cpp:102]: I know this PR didn't introduce the setBaudrate code, but upon examination, it looks the implementation could be incompatible with existing configurations with baudrates higher than B921600 (B1000000 and higher). The RTPS implementation had support for 1M-4M, so maybe worth adding those base in here in this pass?

File: src/modules/microdds_client/dds_topics.yaml (hunk 5 above)
Comment (dagar) [reply src/modules/microdds_client/dds_topics.yaml:3]: Not really, it all goes back to things like IMU_GYRO_RATEMAX (and the actual IMU), IMU_INTEG_RATE, etc, so the the range in output data rate can be enormous for certain topics. 

In general though I'm going to take a closer look at optimizing some of these higher rate topics, being a bit more careful with the size and maybe do more to reduce the publication rate when there's little to no change.

File: src/modules/microdds_client/microdds_client.cpp (hunk 6 above)
Comment (dagar) [reply src/modules/microdds_client/microdds_client.cpp:102]: Interesting, I didn't realize is was that high. Let's get it updated.

File: src/modules/microdds_client/microdds_client.cpp (hunk 1 above)
Comment (dagar) [reply src/modules/microdds_client/microdds_client.cpp:496]: Thanks, fixed.

**Comment by alvinsunyixiao** [comment]
Is there any timeline on when this transition will happen?

**Comment by Jaeyoung-Lim** [comment]
I have tried running this branch with the `micro-ros-agent` and it works great.

However, I am somehow struggling to subscribe to topics in my subscriber

There were also a few strange things observed along the way
 - The rqt topic monitor seems to not be able to display the topic names properly
image


@dsix-ls2n @TSC21 Any ideas why this could be?

**Comment by dsix-ls2n** [comment]
Hello, @Jaeyoung-Lim I have the same display in the topic monitor (Humble, ubuntu 22.04). However, I have no issue subscribing to the topics. The two issues seem unrelated.

Maybe an issue with the QoS profile? Personally, I use this settings in the subscriber
QoSProfile(
            reliability=QoSReliabilityPolicy.BEST_EFFORT,
            history=QoSHistoryPolicy.KEEP_LAST,
            depth=1
        )

**Comment by dsix-ls2n** [comment]
Also I use the master of micro_ros_agent not the snap, I don't know if it is up to date

**Comment by junwoo091400** [comment]
@Jaeyoung-Lim Could you share the details on how you used the `micro-ros-agent`? Or @dsix-ls2n

That would be helpful for ones without much ROS knowledge :thinking: (like me). In the meantime I will try to figure it out

**Comment by Jaeyoung-Lim** [comment]
@junwoo091400 You can check the readme of https://github.com/Jaeyoung-Lim/px4-offboard

File: src/modules/microdds_client/dds_topics.yaml [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    type: px4_msgs::msg::VehicleTrajectoryWaypoint
+
+  - topic: /fmu/in/onboard_computer_status
+    type: px4_msgs::msg::OnboardComputerStatus
+
+  - topic: /fmu/in/trajectory_bezier
+    type: px4_msgs::msg::TrajectoryBezier
+
+  - topic: /fmu/in/vehicle_trajectory_bezier
+    type: px4_msgs::msg::VehicleTrajectoryBezier
```
Comment (Jaeyoung-Lim) [inline src/modules/microdds_client/dds_topics.yaml:96]: Could we add `px4_msgs::msg::VehicleRatesSetpoint` and `px4_msgs::msg::VehicleAttitudeSetpoint` to the subscription topic list?

**Review by Jaeyoung-Lim — APPROVED** [review]
Thanks! 

This has worked great for me using ROS2 with https://github.com/Jaeyoung-Lim/px4-offboard with SITL

Minor comments to get more controller related topics to be exposed

File: src/modules/microdds_client/dds_topics.yaml (hunk 7 above)
Comment (dagar) [reply src/modules/microdds_client/dds_topics.yaml:83]: Done.

**Review by Jaeyoung-Lim — DISMISSED** [review]

**Comment by Jaeyoung-Lim** [comment]
@dagar Are the CI failures related to this PR?

**Review by Jaeyoung-Lim — APPROVED** [review]

**Comment by junwoo091400** [comment]
The `standard_vtol` test failing is due to `[   9.027|gzserver  ] recvfrom error: Connection reset by peer`

image

**Comment by junwoo091400** [comment]
Oh seems like ur commit already contains the isFinite function implementation for Vectors. That would close https://github.com/PX4/PX4-Autopilot/pull/19569 🔥

**Comment by dagar** [comment]
There's quite a lot more to do here, but this is now in a usable tested state. With this out of the way it will unblock more incremental ROS2 unification work, so we might as well get on with it.

Thanks everyone for the review, testing, and discussion.

**Comment by hamishwillee** [comment]
There is a docs PR for this just started. It's a change that touches a lot of places - will need some expert support to tidy the mess.

Confirmation of this would help: https://github.com/PX4/PX4-user_guide/pull/2094#issuecomment-1284879342

That gives high level context, then we will need to look at detail on all the other docs you can see touched by this in that PR.

**Comment by dsix-ls2n** [comment]
> There's quite a lot more to do here, but this is now in a usable tested state. With this out of the way it will unblock more incremental ROS2 unification work, so we might as well get on with it.
> 
> Thanks everyone for the review, testing, and discussion.

Great, thanks for the work!

**Pull request merged**